### PR TITLE
vsphere-support

### PR DESCRIPTION
**Pull request by imkin**
Fixing issues vsphere support.

### CODE REVIEW

**Comment by imkin** [comment]
I updated the email and trying to see if this helps.

I signed it

**Comment by imkin** [comment]
I signed it!

**Comment by mikedanese** [comment]
@imkin Did you sign an individual CLA or a corporate CLA? I see neither an individual or a corporate.

**Comment by mikedanese** [comment]
@k8s-bot ok to test

File: docs/getting-started-guides/vsphere.md [hunk 1]
```diff
@@ -99,7 +102,7 @@ parameters. The guest login for the image that you imported is `kube:kube`.
 ### Starting a cluster
 
 Now, let's continue with deploying Kubernetes.
-This process takes about ~10 minutes.
+This process takes about ~20-30 minutes depending on your network.
```
Comment (mikedanese) [inline docs/getting-started-guides/vsphere.md]: What part takes this long?

File: cluster/saltbase/salt/docker/init.sls [hunk 2]
```diff
@@ -48,7 +48,6 @@ docker:
 
 {% endif %}
 
-{% else %}
```
Comment (mikedanese) [inline cluster/saltbase/salt/docker/init.sls]: This is the else for debian. This should not be removed

File: cluster/saltbase/salt/kubelet/default [hunk 3]
```diff
@@ -3,7 +3,7 @@
   {% set daemon_args = "" -%}
 {% endif -%}
 
-{% if grains.api_servers is defined -%}
+{% if grains.api_servers is defined -%}`
```
Comment (mikedanese) [inline cluster/saltbase/salt/kubelet/default]: what is ` for

**Comment by mikedanese** [comment]
This change breaks cluster turnup on GCE

File: cluster/saltbase/salt/docker/docker-defaults [hunk 4]
```diff
@@ -1,7 +1,8 @@
 {% set grains_opts = grains.get('docker_opts', '') -%}
-{% set e2e_opts = '' -%}
-{% if pillar.get('e2e_storage_test_environment', '').lower() == 'true' -%}
-{% set e2e_opts = '-s devicemapper' -%}
+{% set e2e_opts = '' %}
+{% set e2e_env = (pillar.get('e2e_storage_test_environment', ''))|string() %}
+{% if e2e_env.lower() == 'true' -%}
```
Comment (zmerlynn) [inline cluster/saltbase/salt/docker/docker-defaults]: line 3-4 seem like a complicated rewrite of the old line 3?

File: cluster/saltbase/salt/kubelet/default [hunk 5]
```diff
@@ -54,7 +54,7 @@
 
 {% set manifest_url = "" -%}
 {% set manifest_url_header = "" -%}
-{% if pillar.get('enable_manifest_url', '').lower() == 'true' %}
+{% if pillar.get('enable_manifest_url', False) == True %}
```
Comment (zmerlynn) [inline cluster/saltbase/salt/kubelet/default]: This is another string-y thing and needs to be left this way. (Yes, it sucks.)

Comment (zmerlynn) [inline cluster/saltbase/salt/kubelet/default]: (It gets pulled into the pillars dict as a string in all the cloud providers.)

File: cluster/saltbase/salt/top.sls [hunk 6]
```diff
@@ -51,14 +51,14 @@ base:
     - kube-controller-manager
     - kube-scheduler
     - supervisor
-{% if grains['cloud'] is defined and not grains.cloud in [ 'aws', 'gce', 'vagrant' ] %}
+{% if grains['cloud'] is defined and not grains.cloud in [ 'aws', 'gce', 'vagrant'] %}
```
Comment (zmerlynn) [inline cluster/saltbase/salt/top.sls]: Unnecessary whitespace change.

File: cluster/saltbase/salt/top.sls [hunk 7]
```diff
@@ -51,14 +51,14 @@ base:
...
     - nginx
 {% endif %}
     - cadvisor
     - kube-client-tools
     - kube-master-addons
     - kube-admission-controls
-{% if pillar.get('enable_node_logging', '').lower() == 'true' and pillar['logging_destination'] is defined %}
+{% if pillar.get('enable_node_logging','').lower() == 'true' and pillar['logging_destination'] is defined %}
```
Comment (zmerlynn) [inline cluster/saltbase/salt/top.sls]: ditto

File: cluster/saltbase/salt/docker/docker-defaults (hunk 4 above)
Comment (imkin) [inline cluster/saltbase/salt/docker/docker-defaults]: This would have made sure that boolean and string both would safely evaluate to the == 'true'. 

Reverting.

File: cluster/saltbase/salt/kubelet/default (hunk 5 above)
Comment (imkin) [inline cluster/saltbase/salt/kubelet/default]: reverted

File: cluster/saltbase/salt/kubelet/default [hunk 8]
```diff
@@ -65,7 +65,7 @@
 
 {% set cluster_dns = "" %}
 {% set cluster_domain = "" %}
-{% if pillar.get('enable_cluster_dns', '').lower() == 'true' %}
+{% if pillar.get('enable_cluster_dns', False) == True %}
```
Comment (imkin) [inline cluster/saltbase/salt/kubelet/default]: reverted

File: docs/getting-started-guides/vsphere.md (hunk 1 above)
Comment (imkin) [inline docs/getting-started-guides/vsphere.md]: salt packages installation [apt-update takes a lot of time].

File: cluster/saltbase/salt/top.sls (hunk 7 above)
Comment (imkin) [inline cluster/saltbase/salt/top.sls]: reverted.

**Comment by zmerlynn** [comment]
@k8s-bot test this

**Comment by imkin** [comment]
@mikedanese 
@zmerlynn I see a GCE e2e test is failing. My changes do not touch GCE portion. Any suggestions on where the failure is?

**Comment by zmerlynn** [comment]
@k8s-bot test this

**Comment by mikedanese** [comment]
@k8s-bot unit test this please

**Comment by imkin** [comment]
@mikedanese It seems to be 3 hours and the e2e seems to have passed. unit tests seems to be hung somewhere.

**Comment by mikedanese** [comment]
Ya looks like a flake

@k8s-bot unit test this please

**Comment by frapposelli** [comment]
@mikedanese I had an exchange with our legal dept. and we do have a corporate CLA signed as VMware, maybe the CLA has changed since then and must be signed again?

**Comment by frapposelli** [comment]
@mikedanese we should be good to go with the CLA, we re-signed the corporate CLA and @imkin is in the appropriate google group.

**Comment by mikedanese** [comment]
@frapposelli Thanks for resending! I can confirm that we've recieved the CLA and we should be good to go. Please give it a day to be processed.

**Comment by mikedanese** [comment]
Lawyers say this is good to go.

File: cluster/vsphere/config-default.sh [hunk 9]
```diff
@@ -45,7 +45,7 @@ ELASTICSEARCH_LOGGING_REPLICAS=1
 #   none     - No cluster monitoring setup 
 #   influxdb - Heapster, InfluxDB, and Grafana 
 #   google   - Heapster, Google Cloud Monitoring, and Google Cloud Logging
-ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-influxdb}"
+ENABLE_CLUSTER_MONITORING="${KUBE_ENABLE_CLUSTER_MONITORING:-none}"
```
Comment (mikedanese) [inline cluster/vsphere/config-default.sh]: might want to note somewhere that horizontal pod autoscaling will not work without heapster.

**Comment by mikedanese** [comment]
@imkin This looks good to me. Is there any way we could setup some CI that runs some conformance tests so we can see when this breaks. We have this for GCE, AWS, and mesos integrations.

**Comment by frapposelli** [comment]
@mikedanese we're setting up a public CI infrastructure, it's going to be ready soon

**Comment by veryhumble** [comment]
Hi, I just tried this and it's not quite working yet.

I downloaded the latest Release v1.2.0-alpha.6 and patched the files myself. The master and all nodes are running but no containers get deployed due to some kind of kernel error:

```
Jan 21 09:58:06 kubernetes-master systemd[1]: Failed to set cpu.cfs_period_us on /system.slice/docker-dd431ac2a8967d34ce5e87c8f3bcd02bbc84554f27f31f6b537626e36603683a.scope: Permission denied
Jan 21 09:58:06 kubernetes-master systemd[1]: Failed to set cpu.cfs_quota_us on /system.slice/docker-dd431ac2a8967d34ce5e87c8f3bcd02bbc84554f27f31f6b537626e36603683a.scope: Permission denied
```

What I found was this: https://lists.debian.org/debian-kernel/2015/10/msg00313.html
So I updated all packages using 
```
apt-get dist-upgrade
```

And the kernel version now is:
```
3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u3 (2016-01-17)
```

Still, no containers get deployed. Am I missing something?

Kind regards.

**Comment by alex-mohr** [comment]
@k8s-bot test this
@k8s-bot unit test this

**Comment by imkin** [comment]
@veryhumble 

We do currently have an issue https://github.com/kubernetes/kubernetes/pull/19354
But it is totally different from what you are reporting.

This is the version of the kernel that is part of the kube.vmdk that is mentioned in https://github.com/cloudnativeapps/kubernetes/blob/vsphere-support-1/docs/getting-started-guides/vsphere.md

"3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u2"


Please use the vmdk [ and the instructions ] mentioned in the
https://github.com/cloudnativeapps/kubernetes/blob/vsphere-support-1/docs/getting-started-guides/vsphere.md

and use the tree https://github.com/cloudnativeapps/kubernetes/tree/vsphere-support-1 and apply the patch https://github.com/kubernetes/kubernetes/pull/19354 and things will work.

File: cluster/vsphere/config-default.sh (hunk 9 above)
Comment (imkin) [inline cluster/vsphere/config-default.sh]: I wanted to add this incrementally. SInce I had more time I added influxdb back and tested it.

**Comment by imkin** [comment]
@mikedanese I have satisfied your comment and all tests have passed.

**Comment by mikedanese** [comment]
Still LGTM

**Comment by mikedanese** [comment]
@k8s-bot e2e test this please
